### PR TITLE
Ilp structure integration

### PR DESCRIPTION
**Pull request by Wolokin**
Not very happy with the owa impl, I found it quite inconvenient to actually add owa_vector as parameter somewhere, so it's in closure now. I've thought of several solutions to fix this, none of which were satisfying to me :(

### CODE REVIEW

File: votepy/algorithms/base_ilp.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+class ILP(BaseAlgorithm):
+
+    def __init__(self, solver=Gurobi):
+        """Base class for ILP algorithms"""
+        super().__init__()
+        self.solver = solver
+
+    def prepare(self, define_model):
+        self.define_model = define_model
+        super().prepare()
```
Comment (Janix4000) [inline votepy/algorithms/base_ilp.py]: You can add description/example/type hint of the `define_model`. At the first glance I thought it is a string descripton of the ILP.

**Review by Janix4000 — COMMENTED** [review]
Our library structure does not support such cases like your `custom_cc_ilp`. Maybe we should discuss these scenarios?

File: votepy/algorithms/base_ilp.py (hunk 1 above)
Comment (Wolokin) [reply votepy/algorithms/base_ilp.py]: Yeah, I forgot about that, sorry

File: votepy/algorithms/base_ilp.py [hunk 2]
```diff
@@ -1,22 +1,25 @@
...
-    def __init__(self, solver=Gurobi):
+    def __init__(self, solver: solver_t = Gurobi):
         """Base class for ILP algorithms"""
         super().__init__()
         self.solver = solver
 
-    def prepare(self, define_model):
+    def prepare(self, define_model: Callable[
+        [OrdinalElection, int, solver_t],
+            model_t]) -> None:
```
Comment (Janix4000) [inline votepy/algorithms/base_ilp.py]:
```suggestion
    committee_size_t = int
    def prepare(self, define_model: Callable[
        [OrdinalElection, committee_size_t, solver_t],
            model_t]) -> None:
```

**Review by Janix4000 — CHANGES_REQUESTED** [review]

**Comment by Wolokin** [comment]
[Rewrite all ILP rules to conform to the new library structure](https://trello.com/c/DOa35cpq/76-rewrite-all-ilp-rules-to-conform-to-the-new-library-structure)

**Review by Janix4000 — APPROVED** [review]
`ILP.prepare` should be better documented (with some examples) but it is good enough.

**Pull request merged**